### ISSUE TITLE
500 Server Error due to display of reservation questions

### ISSUE THREAD

**Issue by jat255** [usnistgov/NEMO#145]
We recently noticed an issue with certain reservations returning a server error when trying to view the "event details":

![image](https://user-images.githubusercontent.com/1278301/233743816-d88a7385-15f0-4860-8536-5c20b911ac32.png)

![image](https://user-images.githubusercontent.com/1278301/233743879-4faf2e6c-a24d-461d-ad14-4f360b161372.png)

After some debugging, it appears this is occurring due to groups within the reservation questions, and what happens if a value defined for some items of the group, but not others (we noticed this with the periodic table question).

Here's the structure we're using for questions:

 ```json
[
  {
    "type": "textbox",
    "name": "project_id",
    "title": "Project ID",
    "title_html": "<strong>Project ID: </strong>",
    "max-width": "350",
  },
  {
    "type": "textbox",
    "name": "experiment_title",
    "title": "Title of Experiment",
    "title_html": "<strong>Title of Experiment: </strong>",
    "max-width": "350",
    "required": true
  },
  {
    "type": "textarea",
    "name": "experiment_purpose",
    "title": "Experiment Purpose",
    "title_html": "<strong>Experiment Purpose: </strong>",
    "max-width": "350",
    "rows": "5",
    "required": true
  },
  {
    "type": "radio",
    "title": "Agree to curation",
    "title_html": "<strong>Curate this session's data: </strong>",
    "choices": ["Agree", "Disagree"],
    "name": "data_consent",
    "max-width": "350",
    "required": false,
    "default_choice": "Agree"
  },
{
    "type": "group",
    "title": "Sample information",
    "title_html": "<strong>Sample information: </strong><p>Enter information for at least 1 sample below, and click the \"Add\" button to add up to 5 samples.</p>",
    "name": "sample_group",
    "max_number": 5,
    "questions":
    [
      {
        "type": "textbox",
        "name": "sample_name",
        "max-width": "350",
        "required": true,
        "title": "Sample Name / PID",
        "title_html": "<em>Sample Name / PID: </em>",
        "help": "Write in a sample name, or paste a persistent identifier (PID) for your specimen from one of the LIMS/repositories above."
      },
      {
        "type": "radio",
        "title": "Sample or PID?",
        "title_html": "<em>Is this a sample name or a PID? </em>",
        "choices": ["Sample Name", "PID"],
        "name": "sample_or_pid",
        "required": true,
        "default_choice": null
      },
      {
        "type": "textarea",
        "name": "sample_details",
        "title": "Sample Details",
        "title_html": "<em>Sample Details: </em>",
        "max-width": "350",
        "rows": "3",
        "required": false
      },
      {
        "type": "periodic-table",
        "title": "Sample elements",
        "title_html": "<em>What elements do you expect to be in your specimen? </em>",
        "name": "periodic_table",
        "required": false,
        "collapsible": true,
        "collapsed": false
       }
    ]
  }
]
```

And here's an example of question data on a reservation that causes an error:

```json
{
	"project_id": {
		"type": "textbox",
		"name": "project_id",
		"title": "Project ID",
		"title_html": "<strong>Project ID: </strong>",
		"max-width": "350",
		"user_input": "test"
	},
	"experiment_title": {
		"type": "textbox",
		"name": "experiment_title",
		"title": "Title of Experiment",
		"title_html": "<strong>Title of Experiment: </strong>",
		"max-width": "350",
		"required": true,
		"user_input": "test"
	},
	"experiment_purpose": {
		"type": "textarea",
		"name": "experiment_purpose",
		"title": "Experiment Purpose",
		"title_html": "<strong>Experiment Purpose: </strong>",
		"max-width": "350",
		"rows": "5",
		"required": true,
		"user_input": "test"
	},
	"data_consent": {
		"type": "radio",
		"title": "Agree to NexusLIMS curation",
		"title_html": "<strong>Curate this session's data: </strong>",
		"choices": [
			"Agree",
			"Disagree"
		],
		"name": "data_consent",
		"max-width": "350",
		"required": false,
		"default_choice": "Agree",
		"user_input": "Agree"
	},
	"sample_group": {
		"type": "group",
		"title": "Sample information",
		"title_html": "<strong>Sample information: </strong>",
		"name": "sample_group",
		"max_number": 5,
		"questions": [
			{
				"type": "textbox",
				"name": "sample_name",
				"max-width": "350",
				"required": true,
				"title": "Sample Name / PID",
				"title_html": "<em>Sample Name / PID: </em>",
				"help": "Write in a sample name, or paste a persistent identifier (PID) for your specimen from one of the LIMS/repositories above."
			},
			{
				"type": "radio",
				"title": "Sample or PID?",
				"title_html": "<em>Is this a sample name or a PID? </em>",
				"choices": [
					"Sample Name",
					"PID"
				],
				"name": "sample_or_pid",
				"required": true,
				"default_choice": null
			},
			{
				"type": "textarea",
				"name": "sample_details",
				"title": "Sample Details",
				"title_html": "<em>Sample Details: </em>",
				"max-width": "350",
				"rows": "3",
				"required": false
			},
			{
				"type": "periodic-table",
				"title": "Sample elements",
				"title_html": "<em>What elements do you expect to be in your specimen? </em>",
				"name": "periodic_table",
				"required": false,
				"collapsible": true,
				"collapsed": false
			}
		],
		"user_input": {
			"0": {
				"sample_name": "test1",
				"sample_or_pid": "Sample Name",
				"sample_details": "test",
				"periodic_table": [
					"Ru",
					"Ag",
					"Cd"
				]
			},
			"1": {
				"sample_name": "test2",
				"sample_or_pid": "Sample Name",
				"sample_details": "test2"
			}
		}
	}
}
```

In particular, the issue comes from the "sample" group:

```json
"user_input": {
        "0": {
		"sample_name": "test1",
		"sample_or_pid": "Sample Name",
		"sample_details": "test",
		"periodic_table": [
			"Ru",
			"Ag",
			"Cd"
		]
	},
	"1": {
		"sample_name": "test2",
		"sample_or_pid": "Sample Name",
		"sample_details": "test2"
	}
}
```

If I change this to the following by adding an empty "periodic_table" response:

```json
"user_input": {
        "0": {
		"sample_name": "test1",
		"sample_or_pid": "Sample Name",
		"sample_details": "test",
		"periodic_table": [
			"Ru",
			"Ag",
			"Cd"
		]
	},
	"1": {
		"sample_name": "test2",
		"sample_or_pid": "Sample Name",
		"sample_details": "test2",
		"periodic_table": [ ]
	}
}
```

Then there is no longer an error:

![image](https://user-images.githubusercontent.com/1278301/233744553-67c0f08b-06a0-432b-99a2-4f07a43a00aa.png)

I haven't debugged in the application code itself, but I suspect the error is coming from something I wrote (sorry!) in #97. The fastest workaround solution I see would be to make the periodic table plugin return an empty list if no elements are chosen. The proper fix would probably be to change `NEMO/templates/event_details/reservation_details.html` so it can handle cases like this where some elements in a group are not defined.

**Comment by rptmat57** [comment]
Thanks for the very detailed bug report!
Fixed in 4.5.4

**Comment by jat255** [comment]
Okay, that might break the record for the fastest bugfix release I've ever seen... :exploding_head:
